storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java

```diff
@@ -16,6 +16,9 @@
  */
 package org.apache.kafka.storage.internals.log;
 
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
```

Comment:
We introduced a new dependency caffeine here. Could you explain why caffeine is chosen and how stable is caffeine? The doc for caffeine mentions the use of weak references. A few years back, we avoided the usage of weak references in a PR because of the poor GC behavior. Have we done any experiments to understand the GC impact?

Reply:
I think this was already there for the Scala code.

Reply:
Caffeine is added in this [PR](https://github.com/apache/kafka/pull/13850) to improve concurrent read performance. We chose it since it has good performance and is also adopted by other big projects, like Cassandra, HBase etc. About the GC issues, I don't think we have discussed about it. cc @divijvaidya , since you've done some research on Caffeine, do you have any comment to the GC issue?

Reply:
Hey Jun

Caffeine is being used for many years in other latency/memory sensitive projects such as Solr, Cassandra.

Solr: It is being used in Apache Solr since 2019[1]. From their release doc[2]:
> Users are encouraged to transition their cache configurations to use org.apache.solr.search.CaffeineCache as soon as feasible. 

Cassandra: It is being used in Cassandra since 2017 for performance critical caches.

Now, about the weak references, please note that weak & soft references are optional [4], and we don't use weak/soft references in this implementation.. The keys and values are by default stored with strong references. You can see this information at the javadoc [4] the `newBuilder()` method we are using. Quoting from there:
> Constructs a new Caffeine instance with default settings, including strong keys, strong values, and no automatic eviction of any kind.

Personally, outside of Kafka, I have used this cache in query execution path of database internals and it hasn't caused any problems. Please let me know if you have further question about it's usage. Happy to explain more!


[1] https://issues.apache.org/jira/browse/SOLR-8241
[2] https://solr.apache.org/guide/8_4/solr-upgrade-notes.html#Caches
[3] https://issues.apache.org/jira/browse/CASSANDRA-10855
[4] https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/2.0.3/com/github/benmanes/caffeine/cache/Caffeine.html#newBuilder--

Reply:
Thanks for the explanation, Divij. This change seems fine to me then.